modules/renter/contractor/downloader.go

```diff
@@ -74,78 +92,53 @@ func (hd *hostDownloader) Sector(root crypto.Hash) ([]byte, error) {
 	if hd.invalid {
 		return nil, errInvalidDownloader
 	}
-	contract, sector, err := hd.downloader.Sector(root)
+
+	// Download the sector.
+	_, sector, err := hd.downloader.Sector(root)
 	if err != nil {
 		return nil, err
 	}
-
-	hd.contractor.mu.Lock()
-	hd.contractor.contracts[contract.ID] = contract
-	hd.contractor.persist.update(updateDownloadRevision{
-		NewRevisionTxn:      contract.LastRevisionTxn,
-		NewDownloadSpending: contract.DownloadSpending,
-	})
-	hd.contractor.mu.Unlock()
-
 	return sector, nil
 }
 
-// Close cleanly terminates the download loop with the host and closes the
-// connection.
-func (hd *hostDownloader) Close() error {
-	hd.mu.Lock()
-	defer hd.mu.Unlock()
-	hd.clients--
-	// Close is a no-op if invalidate has been called, or if there are other
-	// clients still using the hostDownloader.
-	if hd.invalid || hd.clients > 0 {
-		return nil
-	}
-	hd.invalid = true
-	hd.contractor.mu.Lock()
-	delete(hd.contractor.downloaders, hd.contractID)
-	delete(hd.contractor.revising, hd.contractID)
-	hd.contractor.mu.Unlock()
-	return hd.downloader.Close()
-}
-
 // Downloader returns a Downloader object that can be used to download sectors
 // from a host.
 func (c *Contractor) Downloader(id types.FileContractID, cancel <-chan struct{}) (_ Downloader, err error) {
 	id = c.ResolveID(id)
 	c.mu.RLock()
 	cachedDownloader, haveDownloader := c.downloaders[id]
 	height := c.blockHeight
-	contract, haveContract := c.contracts[id]
 	renewing := c.renewing[id]
 	c.mu.RUnlock()
-
 	if renewing {
 		return nil, errors.New("currently renewing that contract")
-	}
-
-	if haveDownloader {
+	} else if haveDownloader {
 		// increment number of clients and return
 		cachedDownloader.mu.Lock()
 		cachedDownloader.clients++
 		cachedDownloader.mu.Unlock()
 		return cachedDownloader, nil
 	}
 
-	host, haveHost := c.hdb.Host(contract.HostPublicKey)
+	// Fetch the contract and host.
+	contract, haveContract := c.contracts.View(id)
 	if !haveContract {
 		return nil, errors.New("no record of that contract")
-	} else if height > contract.EndHeight() {
+	}
+	host, haveHost := c.hdb.Host(contract.HostPublicKey)
+	if height > contract.EndHeight {
 		return nil, errors.New("contract has already ended")
 	} else if !haveHost {
 		return nil, errors.New("no record of that host")
 	} else if host.DownloadBandwidthPrice.Cmp(maxDownloadPrice) > 0 {
 		return nil, errTooExpensive
 	}
-	// Update the contract to the most recent net address for the host.
-	contract.NetAddress = host.NetAddress
 
-	// acquire revising lock
+	// Acquire the revising lock for the contract, which excludes other threads
+	// from interacting with the contract.
+	//
+	// TODO: Because we have another layer of contract safety via the
+	// contractset, do we need the revising lock anymore?
```

Comment:
Need to double check everything that the revising lock is protecting, but I believe that the contractset now provides all the safety that you need, and that we can drop the revising lock.